### ISSUE TITLE
RELP SSL module Version 8.1908.0-1.1 SSL certpath inconsistent config behaviour

### ISSUE THREAD

**Issue by Bassbauer** [rsyslog/rsyslog#3881]
### Expected behavior
global(
  workDirectory="/rsyslogd"
#  defaultNetstreamDriverCAFile="/rsyslogd/certs/ca.pem"
#  defaultNetstreamDriverCertFile="/rsyslogd/certs/cert.pem"
#  defaultNetstreamDriverKeyFile="/rsyslogd/certs/cert.key"
  maxMessageSize="64k"
)
input(type="imrelp" port="7805"
      ruleset="rule_7805"
      tls="on"
      tls.cacert="certs/ca.pem"
      tls.mycert="certs/cert.pem"
      tls.myprivkey="certs/cert.key"
      tls.authmode="name"
      tls.permittedPeer="*.test.com"
)

should honor indirect paths and load the gnutls / openssl module for imrelp.

### Actual behavior
the imrelp module does not honor the indirect path, however when specifying the path with absolute values it seems to work. When specifiying the defaultNetstreamDriver... values all other inputs seem to work.

### Steps to reproduce the behavior
try to configure imrelp module with SSL with relative paths
compare behaviour when one input is specified with absolute path and all others without, all other modules suddenly work as well.

If this is the expected behaviour (only one certificate can be used for all inputs) then this should be documented.

### Environment
rsyslogd  8.1908.0 (aka 2019.08) compiled with:
        PLATFORM:                               x86_64-suse-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Config file:                            /etc/rsyslog.conf
        PID file:                               /var/run/rsyslogd.pid
        Number of Bits in RainerScript integers: 64

See https://www.rsyslog.com for more information.

**Comment by Bassbauer** [comment]
It seems that there is some general problem with relative path evaluation in newer versions, maybe that is all related:

in work directory
/rsyslog

in version 8.1908.0:
/usr/sbin/rsyslogd -n -f rsyslogd.conf -i pid/rsyslogd.pid
Does not work.

in version 8.24.0
/usr/sbin/rsyslogd -n -f rsyslogd.conf -i pid/rsyslogd.pid
Does work.

I would assume that all paths should work relative to 

global(
 workDirecroty="/rsyslog"
)

I can compensate by using absolute paths. Maybe iam overlooking something here?

**Comment by davidelang** [comment]
relative directories would be relative to CWD for rsyslog, not relative to the 
work directory

the work directory is just where rsyslog stores it's admin files (queue files, 
etc)

David Lang

**Comment by Bassbauer** [comment]
Hi David,

i have tested the same with CWD and the behaviour is still the same, to use the CWD is logical and also expected behaviour. Interestingly in older version it is working correctly, i guess thats some kind of regression.

Thanks,
Frank

**Comment by rgerhards** [comment]
I don't think we have a regression here.  The default working directory is the root (`/`) and this really has not changed. Which version it did "work" with? Can you create a debug log with that version?

**Comment by Bassbauer** [comment]
A debug log is problematic since it contains to much information about the environment, if it is really necessary i can anonymise it. But in generall is anyone able to configure a certificate in the imrelp module with a relative path?

Additionally the interference of defaultNetstreamDriverCAFile* is hard to understand since all modules will use that default even if a specific 

tls.cacert="certs/ca.pem"
tls.mycert="certs/cert.pem"
tls.myprivkey="certs/cert.key"

is set (and those files cant be found).

To make sure we have the same understanding:

CWD is /rsyslog

then

tls.cacert="certs/ca.pem"
tls.mycert="certs/cert.pem"
tls.myprivkey="certs/cert.key"

should work? (if a folder with certs/ exist with those files).

i have migrated a configuration from  8.24.0 -> 8.1908.0 

Regards,

Frank

**Comment by rgerhards** [comment]
I am not sure how you set the working directory. Note that if you do **not** give the -C rsyslogd command line option, rsyslog will do a `chdir("/")` as this is the traditional and expected behavior.

**Comment by Bassbauer** [comment]
that solves the problem rsyslog behaves as it did in the old version and relative paths work. However as the documentation suggest it is "almost never a good idea" to use this feature.

It still seems like a regression to me, what is the planned way to use relative paths in rsyslogd?

Why does the new version needs -C ?

iam starting rsyslogd as a user with user permissions only.

rsyslog@test:~/rsyslogd> /usr/sbin/rsyslogd -dn -f rsyslogd.conf -i pid/rsyslogd.pid -u rsyslogd

works when i use the -C flag.

input(type="imrelp" port="7202"
      ruleset="rule_7202tcp_relp_enc"
      tls="on"
      tls.cacert="certs/ca.pem"
      tls.mycert="certs/cert.pem"
      tls.myprivkey="certs/cert.key"
      tls.authmode="certvalid"
      tls.permittedPeer="*.test.com"
)
ruleset(name="rule_7202tcp_relp_enc"
        queue.type="LinkedList"
        queue.size="100000"
        queue.dequeueBatchSize="4096"
        queue.workerThreads="8"
        queue.filename="queue_7202tcp_relp_enc"
        queue.spoolDirectory="disk_queues"
        queue.maxDiskSpace="10g"
        queue.workerThreadMinimumMessages="10000"
       ) {
     action(
           FileOwner="rsyslog"
           FileGroup="rsyslog"
           FileCreateMode="0660"
           type="omfile" file="spool/7202tcp_relp_enc.log"
           ioBufferSize="64k" flushOnTXEnd="on"
           asyncWriting="on"
           flushInterval="1"
           template="RemoteHostFileFormat")
}

without the -C flag i have to use absolute paths, the -C flag is not necessary in Version 8.24.0.

Could anyone confirm if the issue can be reproduced?

Thanks

**Comment by rgerhards** [comment]
> that solves the problem rsyslog behaves as it did in the old version and relative paths work.

There is a misunderstanding: the handling is absolutely unchanged since -C was introduced several years ago. And before it, rsyslog always did the `chdir("/")`. That means everything is working as expected and as it always had.

Just to be on the safe side, I checked the 8.24.0 source code and the code is indeed exactly the same.

Maybe you used a packaged version with a custom patch that we don't know about.

**Comment by Bassbauer** [comment]
we are using rsyslog on SLES, i can indeed not rule out that the package contains modification. However may i suggest to add a comment to the manual that the use of -C is valid when rsyslog does not run as root? That was the reason i did not consider using it initially.

**Comment by rgerhards** [comment]
Apologies for the very late reply. This may no longer be relevant for the original reporter, but I am adding the answer for people who find this issue later via search. We do not lightly close old issues, and we are working through the backlog rather than closing issues just because of age.

The important conclusion from the thread is that `global(workDirectory="...")` is not a base directory for resolving arbitrary relative paths in configuration. It is the working directory rsyslog uses for internal/admin files such as queues.

Relative paths in configuration are relative to the process current working directory. In normal daemon startup rsyslog traditionally changes directory to `/`; current source still has that behavior. If you intentionally want to preserve the caller's current directory for a special non-root/manual run, use the `-C` command-line option, or better, use absolute paths for certificates and other operational files.

So for RELP/TLS certificate files, the robust configuration is:

```rsyslog
input(type="imrelp"
      port="7805"
      tls="on"
      tls.cacert="/rsyslogd/certs/ca.pem"
      tls.mycert="/rsyslogd/certs/cert.pem"
      tls.myprivkey="/rsyslogd/certs/cert.key"
      tls.authmode="name"
      tls.permittedPeer="*.test.com")
```

I am closing this as answered/expected behavior rather than a current rsyslog bug. If there is still a current-version case where an absolute certificate path is ignored or overridden by a global default unexpectedly, please open a new issue with a minimal reproducer.
